src/cryptography/noscrypt_cipher.cpp

```diff
@@ -0,0 +1,146 @@
+    const std::shared_ptr<const NCPublicKey> remoteKey,
+    const std::string& input
+)
+{
+    NCResult result;
+
+    //Argument exception if the input is empty
+    if (input.empty())
+    {
+        return string();
+    }
+
+    //Safely convert the string to a vector of bytes (allocates and copies, so maybe speed up later)
+    const vector<uint8_t> inputBuffer(input.begin(), input.end());
+
+    result = this->_cipher.setInput(inputBuffer);
+    if (result != NC_SUCCESS)
+    {
+        NC_LOG_ERROR(result);
+        return string();
+    }
+
+    /*
+    * If were in encryption mode a random nonce (iv) must be generated. The size was determined
+    * when the cipher was created and already assigned to the context, so we just need to assign
+    * the random data.
+    *
+    * Keep in mind, this will automatically work for nip44 and nip04, either the
+    * AES iv or the ChaCha nonce.
+    */
+    if (this->_cipher.mode() == NoscryptCipherMode::CIPHER_MODE_ENCRYPT)
+    {
+		//A secure random initialization vector is needed for encryption operations
+		NostrSecureRng::fill(this->_ivBuffer);
+    }
+
+    //Performs the operation (either encryption or decryption) 
+    result = this->_cipher.update(libContext, localKey, remoteKey);
+    if (result != NC_SUCCESS)
+    {
+        NC_LOG_ERROR(result);
+        return string();
+    }
+
+    /*
+    * Time to read the ciper output by getting the size of the output, then creating
+    * a string to store it to
+    */
+
+    NCResult outputSize = this->_cipher.outputSize();
+    if (outputSize <= 0)
+    {
+        NC_LOG_ERROR(outputSize);
+        return string();
+    }
+
+    //Alloc vector for reading input data (maybe only alloc once)
+    vector<uint8_t> output(outputSize);
+
+    result = this->_cipher.readOutput(output);
+    if (result != outputSize)
+    {
+        NC_LOG_ERROR(result);
+        return string();
+    }
+
+    return string(output.begin(), output.end());
+}
+
+string NoscryptCipher::naiveEncodeBase64(const std::string& str)
+{
+    // Compute base64 size and allocate a string buffer of that size.
+    const size_t encodedSize = NoscryptCipher::base64EncodedSize(str.size());
+    
+    auto encodedData = make_unique<uint8_t>(encodedSize);
+
+    // Encode the input string to base64.
+    EVP_EncodeBlock(
+        encodedData.get(),
+        reinterpret_cast<const uint8_t*>(str.c_str()),
+        str.size()
+    );
+
+    // Construct the encoded string from the buffer.
+    return string(
+        reinterpret_cast<char*>(encodedData.get()),
+        encodedSize
+    );
+}
+
+string NoscryptCipher::naiveDecodeBase64(const string& str)
+{
+    // Compute the size of the decoded string and allocate a buffer of that size.
+    const size_t decodedSize = NoscryptCipher::base64DecodedSize(str.size());
+  
+    auto decodedData = make_unique<uint8_t>(decodedSize);
+
+    // Decode the input string from base64.
+    EVP_DecodeBlock(
```

Comment:
I think I want us to abstract the OSSL specific stuff to a separate class (it might have to be an instance instead of static) to encapsulate the basic operations we need such as base64 and sha256 stuff. It also makes it way easier to swap out if we find better backends. Ive also learned openssl can be quite hefty and we may want to allow developers to configure how openssl is used under the hood. 

Considering we already wrapped rng into it's own static class, I'm thinking we move all cryptographic operations to an abstract class. 

I'm going to be doing the same thin with noscrypt as well.

Reply:
I wrote an issue for this in #17.  Creating an OpenSSL wrapper can easily be its own user story.